File: backend/src/test/java/com/back/fairytale/domain/thoughts/service/ThoughtsServiceTest.java
package com.back.fairytale.domain.thoughts.service;

import com.back.fairytale.domain.fairytale.entity.Fairytale;
import com.back.fairytale.domain.fairytale.repository.FairytaleRepository;
import com.back.fairytale.domain.thoughts.dto.ThoughtsRequest;
import com.back.fairytale.domain.thoughts.dto.ThoughtsResponse;
import com.back.fairytale.domain.thoughts.dto.ThoughtsUpdateRequest;
import com.back.fairytale.domain.thoughts.entity.Thoughts;
import com.back.fairytale.domain.thoughts.repository.ThoughtsRepository;
import com.back.fairytale.domain.user.entity.User;
import com.back.fairytale.domain.user.repository.UserRepository;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.time.LocalDateTime;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class ThoughtsServiceTest {

    @Mock
    private ThoughtsRepository thoughtsRepository;

    @Mock
    private UserRepository userRepository;

    @Mock
    private FairytaleRepository fairytaleRepository;

    @InjectMocks
    private ThoughtsService thoughtsService;

    private User testUser;
    private Fairytale testFairytale;
    private Thoughts testThought;
    private ThoughtsRequest testThoughtsRequest;
    private ThoughtsUpdateRequest testThoughtsUpdateRequest;

    @BeforeEach
    void setUp() {
        testUser = User.builder()
                .id(1L)
                .nickname("testUser")
                .build();

        testFairytale = Fairytale.builder()
                .id(100L)
                .build();

        testThoughtsRequest = new ThoughtsRequest(
                testFairytale.getId(), "New Thought Name", "New Thought Content", "New Parent Content");
        testThoughtsUpdateRequest = new ThoughtsUpdateRequest(
                "Updated Thought Name", "Updated Thought Content", "Updated Parent Content");

        // 실제 of 메소드를 사용한다고 가정하고, 그 결과로 생성될 객체를 미리 정의합니다.
        testThought = Thoughts.builder()
                .id(1L)
                .fairytale(testFairytale)
                .user(testUser)
                .name(testThoughtsRequest.name())
                .content(testThoughtsRequest.content())
                .parentContent(testThoughtsRequest.parentContent())
                .createdAt(LocalDateTime.now())
                .updatedAt(LocalDateTime.now())
                .build();
    }

    @Test
    @DisplayName("아이생각 작성 성공")
    void createThoughts_success() {
        // Given
        when(userRepository.findById(testUser.getId())).thenReturn(Optional.of(testUser));
        when(fairytaleRepository.findById(testFairytale.getId())).thenReturn(Optional.of(testFairytale));
        // Thoughts.of()는 static 메소드이므로 Mockito로 직접 모의하기 어렵습니다.
        // 여기서는 save 메소드가 호출될 때 어떤 Thoughts 객체가 저장될지 예상하고, 그 객체를 반환하도록 설정합니다.
        when(thoughtsRepository.save(any(Thoughts.class))).thenReturn(testThought);

        // When
        ThoughtsResponse response = thoughtsService.createThoughts(testThoughtsRequest, testUser.getId());

        // Then
        assertThat(response).isNotNull();
        assertThat(response.id()).isEqualTo(testThought.getId());
        assertThat(response.name()).isEqualTo(testThought.getName());
        assertThat(response.content()).isEqualTo(testThought.getContent());
        // ThoughtsResponse에 nickname 필드가 있다면 추가
        // assertThat(response.nickname()).isEqualTo(testUser.getNickname());
        verify(thoughtsRepository, times(1)).save(any(Thoughts.class));
    }

    @Test
    @DisplayName("아이생각 작성 실패 - 유저를 찾을 수 없음")
    void createThoughts_userNotFound() {
        // Given
        when(userRepository.findById(testUser.getId())).thenReturn(Optional.empty());

        // When & Then
        RuntimeException exception = assertThrows(RuntimeException.class,
                () -> thoughtsService.createThoughts(testThoughtsRequest, testUser.getId()));
        assertThat(exception.getMessage()).contains("유저를 찾을 수 없습니다.");
        verify(thoughtsRepository, never()).save(any(Thoughts.class));
    }

    @Test
    @DisplayName("아이생각 작성 실패 - 동화를 찾을 수 없음")
    void createThoughts_fairytaleNotFound() {
        // Given
        when(userRepository.findById(testUser.getId())).thenReturn(Optional.of(testUser));
        when(fairytaleRepository.findById(testFairytale.getId())).thenReturn(Optional.empty());

        // When & Then
        RuntimeException exception = assertThrows(RuntimeException.class,
                () -> thoughtsService.createThoughts(testThoughtsRequest, testUser.getId()));
        assertThat(exception.getMessage()).contains("동화를 찾을 수 없습니다.");
        verify(thoughtsRepository, never()).save(any(Thoughts.class));
    }

    @Test
    @DisplayName("아이생각 조회 성공")
    void getThoughts_success() {
        // Given
        when(thoughtsRepository.findById(testThought.getId())).thenReturn(Optional.of(testThought));

        // When
        ThoughtsResponse response = thoughtsService.getThoughts(testThought.getId(), testUser.getId());

        // Then
        assertThat(response).isNotNull();
        assertThat(response.id()).isEqualTo(testThought.getId());
        assertThat(response.name()).isEqualTo(testThought.getName());
        assertThat(response.content()).isEqualTo(testThought.getContent());
    }

    @Test
    @DisplayName("아이생각 조회 실패 - 아이생각을 찾을 수 없음")
    void getThoughts_thoughtNotFound() {
        // Given
        when(thoughtsRepository.findById(testThought.getId())).thenReturn(Optional.empty());

        // When & Then
        RuntimeException exception = assertThrows(RuntimeException.class,
                () -> thoughtsService.getThoughts(testThought.getId(), testUser.getId()));
        assertThat(exception.getMessage()).contains("thoughts를 찾을 수 없습니다.");
    }

    @Test
    @DisplayName("아이생각 조회 실패 - 접근 권한 없음")
    void getThoughts_accessDenied() {
        // Given
        User anotherUser = User.builder().id(2L).nickname("anotherUser").build();
        when(thoughtsRepository.findById(testThought.getId())).thenReturn(Optional.of(testThought));

        // When & Then
        RuntimeException exception = assertThrows(RuntimeException.class,
                () -> thoughtsService.getThoughts(testThought.getId(), anotherUser.getId()));
        assertThat(exception.getMessage()).contains("접근 권한이 없습니다.");
    }

    @Test
    @DisplayName("아이생각 수정 성공")
    void updateThoughts_success() {
        // Given
        when(thoughtsRepository.findById(testThought.getId())).thenReturn(Optional.of(testThought));

        // When
        ThoughtsResponse response = thoughtsService.updateThoughts(testThought.getId(), testThoughtsUpdateRequest, testUser.getId());

        // Then
        assertThat(response).isNotNull();
        assertThat(response.name()).isEqualTo(testThoughtsUpdateRequest.name());
        assertThat(response.content()).isEqualTo(testThoughtsUpdateRequest.content());
        assertThat(response.parentContent()).isEqualTo(testThoughtsUpdateRequest.parentContent());
        assertThat(testThought.getName()).isEqualTo(testThoughtsUpdateRequest.name()); // 엔티티의 내용도 업데이트되었는지 확인
    }

    @Test
    @DisplayName("아이생각 수정 실패 - 아이생각을 찾을 수 없음")
    void updateThoughts_thoughtNotFound() {
        // Given
        when(thoughtsRepository.findById(testThought.getId())).thenReturn(Optional.empty());

        // When & Then
        RuntimeException exception = assertThrows(RuntimeException.class,
                () -> thoughtsService.updateThoughts(testThought.getId(), testThoughtsUpdateRequest, testUser.getId()));
        assertThat(exception.getMessage()).contains("thoughts를 찾을 수 없습니다.");
    }

    @Test
    @DisplayName("아이생각 수정 실패 - 접근 권한 없음")
    void updateThoughts_accessDenied() {
        // Given
        User anotherUser = User.builder().id(2L).nickname("anotherUser").build();
        when(thoughtsRepository.findById(testThought.getId())).thenReturn(Optional.of(testThought));

        // When & Then
        RuntimeException exception = assertThrows(RuntimeException.class,
                () -> thoughtsService.updateThoughts(testThought.getId(), testThoughtsUpdateRequest, anotherUser.getId()));
        assertThat(exception.getMessage()).contains("접근 권한이 없습니다.");
    }

    @Test
    @DisplayName("아이생각 삭제 성공")
    void deleteThoughts_success() {
        // Given
        when(thoughtsRepository.findById(testThought.getId())).thenReturn(Optional.of(testThought));
        doNothing().when(thoughtsRepository).delete(any(Thoughts.class));

        // When
        thoughtsService.deleteThoughts(testThought.getId(), testUser.getId());

        // Then
        verify(thoughtsRepository, times(1)).delete(testThought);
    }

    @Test
    @DisplayName("아이생각 삭제 실패 - 아이생각을 찾을 수 없음")
    void deleteThoughts_thoughtNotFound() {
        // Given
        when(thoughtsRepository.findById(testThought.getId())).thenReturn(Optional.empty());

        // When & Then
        RuntimeException exception = assertThrows(RuntimeException.class,
                () -> thoughtsService.deleteThoughts(testThought.getId(), testUser.getId()));
        assertThat(exception.getMessage()).contains("thoughts를 찾을 수 없습니다.");
        verify(thoughtsRepository, never()).delete(any(Thoughts.class));
    }

    @Test
    @DisplayName("아이생각 삭제 실패 - 접근 권한 없음")
    void deleteThoughts_accessDenied() {
        // Given
        User anotherUser = User.builder().id(2L).nickname("anotherUser").build();
        when(thoughtsRepository.findById(testThought.getId())).thenReturn(Optional.of(testThought));

        // When & Then
        RuntimeException exception = assertThrows(RuntimeException.class,
                () -> thoughtsService.deleteThoughts(testThought.getId(), anotherUser.getId()));
        assertThat(exception.getMessage()).contains("접근 권한이 없습니다.");
        verify(thoughtsRepository, never()).delete(any(Thoughts.class));
    }
}

File: backend/src/test/java/com/back/fairytale/domain/thoughts/service/ThoughtsServiceTest.kt
package com.back.fairytale.domain.thoughts.service

import com.back.fairytale.domain.fairytale.entity.Fairytale
import com.back.fairytale.domain.fairytale.repository.FairytaleRepository
import com.back.fairytale.domain.thoughts.dto.ThoughtsRequest
import com.back.fairytale.domain.thoughts.dto.ThoughtsUpdateRequest
import com.back.fairytale.domain.thoughts.entity.Thoughts
import com.back.fairytale.domain.thoughts.repository.ThoughtsRepository
import com.back.fairytale.domain.user.entity.User
import com.back.fairytale.domain.user.repository.UserRepository
import jakarta.persistence.EntityNotFoundException
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.junit.jupiter.MockitoExtension
import org.mockito.kotlin.*
import java.util.*

import org.mockito.junit.jupiter.MockitoSettings
import org.mockito.quality.Strictness

@MockitoSettings(strictness = Strictness.LENIENT)
@ExtendWith(MockitoExtension::class)
class ThoughtsServiceTest {

    @Mock
    private lateinit var thoughtsRepository: ThoughtsRepository

    @Mock
    private lateinit var userRepository: UserRepository

    @Mock
    private lateinit var fairytaleRepository: FairytaleRepository

    @InjectMocks
    private lateinit var thoughtsService: ThoughtsService

    // --- Helper Functions to create Mocks ---
    private fun createMockUser(id: Long, name: String): User {
        return mock {
            on { this.id } doReturn id
            on { this.name } doReturn name // 'userName' NPE 방지를 위해 name 속성 stubbing
            on { this.nickname } doReturn name
        }
    }

    private fun createMockFairytale(id: Long): Fairytale {
        return mock { on { this.id } doReturn id }
    }

    private fun createMockThought(id: Long, user: User, fairytale: Fairytale, content: String = "아이생각 내용"): Thoughts {
        return mock {
            on { this.id } doReturn id
            on { this.user } doReturn user
            on { this.fairytale } doReturn fairytale
            on { this.name } doReturn "아이이름"
            on { this.content } doReturn content
            on { this.parentContent } doReturn "부모생각 내용"
        }
    }

    // --- Test Cases ---

    @Test
    @DisplayName("[성공] 아이생각 작성")
    fun createThoughts_Success() {
        // Given: 각 테스트에 필요한 mock 객체를 테스트 내부에서 생성
        val mockUser = createMockUser(1L, "테스트유저")
        val mockFairytale = createMockFairytale(10L)
        val request = ThoughtsRequest(mockFairytale.id!!, "이름", "내용", "부모내용")
        val savedMockThought = createMockThought(1L, mockUser, mockFairytale)

        given(userRepository.findById(mockUser.id!!)).willReturn(Optional.of(mockUser))
        given(fairytaleRepository.findById(mockFairytale.id!!)).willReturn(Optional.of(mockFairytale))
        given(thoughtsRepository.save(any<Thoughts>())).willReturn(savedMockThought)

        // When
        val response = thoughtsService.createThoughts(request, mockUser.id!!)

        // Then
        assertThat(response.name).isEqualTo(savedMockThought.name)
        assertThat(response.content).isEqualTo(savedMockThought.content)
        verify(thoughtsRepository, times(1)).save(any<Thoughts>())
    }

    @Test
    @DisplayName("[실패] 아이생각 작성 - 유저 없음")
    fun createThoughts_Fail_UserNotFound() {
        // Given
        val mockFairytale = createMockFairytale(10L)
        val request = ThoughtsRequest(mockFairytale.id!!, "이름", "내용", "부모내용")
        given(userRepository.findById(any())).willReturn(Optional.empty())

        // When & Then
        assertThrows<EntityNotFoundException> {
            thoughtsService.createThoughts(request, 99L)
        }
    }

    @Test
    @DisplayName("[성공] 아이생각 조회")
    fun getThoughts_Success() {
        // Given
        val mockUser = createMockUser(1L, "테스트유저")
        val mockFairytale = createMockFairytale(10L)
        val thoughtId = 1L
        val mockThought = createMockThought(thoughtId, mockUser, mockFairytale)

        given(thoughtsRepository.findById(thoughtId)).willReturn(Optional.of(mockThought))

        // When
        val response = thoughtsService.getThoughts(thoughtId, mockUser.id!!)

        // Then
        assertThat(response.id).isEqualTo(thoughtId)
        assertThat(response.name).isEqualTo(mockThought.name)
    }

    @Test
    @DisplayName("[실패] 아이생각 조회 - 작성자가 아님")
    fun getThoughts_Fail_Unauthorized() {
        // Given
        val mockUser = createMockUser(1L, "테스트유저")
        val mockOtherUser = createMockUser(2L, "다른유저")
        val mockFairytale = createMockFairytale(10L)
        val thoughtId = 1L
        val mockThought = createMockThought(thoughtId, mockUser, mockFairytale)

        given(thoughtsRepository.findById(thoughtId)).willReturn(Optional.of(mockThought))

        // When & Then
        assertThrows<IllegalStateException> {
            thoughtsService.getThoughts(thoughtId, mockOtherUser.id!!)
        }
    }

    @Test
    @DisplayName("[성공] 아이생각 수정")
    fun updateThoughts_Success() {
        // Given
        val mockUser = createMockUser(1L, "테스트유저")
        val mockFairytale = createMockFairytale(10L)
        val thoughtId = 1L
        val request = ThoughtsUpdateRequest("수정된 이름", "수정된 내용", "수정된 부모내용")
        val mockThought = createMockThought(thoughtId, mockUser, mockFairytale)

        given(thoughtsRepository.findById(thoughtId)).willReturn(Optional.of(mockThought))

        // When
        thoughtsService.updateThoughts(thoughtId, request, mockUser.id!!)

        // Then
        verify(mockThought, times(1)).update(request.name, request.content, request.parentContent)
    }

    @Test
    @DisplayName("[성공] 아이생각 삭제")
    fun deleteThoughts_Success() {
        // Given
        val mockUser = createMockUser(1L, "테스트유저")
        val mockFairytale = createMockFairytale(10L)
        val thoughtId = 1L
        val mockThought = createMockThought(thoughtId, mockUser, mockFairytale)

        given(thoughtsRepository.findById(thoughtId)).willReturn(Optional.of(mockThought))

        // When
        thoughtsService.deleteThoughts(thoughtId, mockUser.id!!)

        // Then
        verify(thoughtsRepository, times(1)).delete(mockThought)
    }
    
    @Test
    @DisplayName("[성공] 동화 ID로 아이생각 조회")
    fun getThoughtsByFairytaleId_Success() {
        // Given
        val mockUser = createMockUser(1L, "테스트유저")
        val mockFairytale = createMockFairytale(10L)
        val thoughtId = 1L
        val mockThought = createMockThought(thoughtId, mockUser, mockFairytale)

        given(thoughtsRepository.findByFairytaleIdAndUserId(mockFairytale.id!!, mockUser.id!!)).willReturn(mockThought)

        // When
        val response = thoughtsService.getThoughtsByFairytaleId(mockFairytale.id!!, mockUser.id!!)

        // Then
        assertThat(response.id).isEqualTo(thoughtId)
        assertThat(response.name).isEqualTo(mockThought.name)
    }

    @Test
    @DisplayName("[실패] 동화 ID로 아이생각 조회 - 아이생각 없음")
    fun getThoughtsByFairytaleId_Fail_NotFound() {
        // Given
        val mockUser = createMockUser(1L, "테스트유저")
        val mockFairytale = createMockFairytale(10L)

        given(thoughtsRepository.findByFairytaleIdAndUserId(mockFairytale.id!!, mockUser.id!!)).willReturn(null)

        // When & Then
        assertThrows<EntityNotFoundException> {
            thoughtsService.getThoughtsByFairytaleId(mockFairytale.id!!, mockUser.id!!)
        }
    }
}
